### PR TITLE
Set DISCOVERY_SERVICE to use local DNS instead of fully qualified DNS…

### LINKED ISSUE #6741: [incubator/elasticsearch] not enough masters
**Is this a request for help?**:

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
 BUG REPORT

**Version of Helm and Kubernetes**:
helm version
Client: &version.Version{SemVer:"v2.7.2", GitCommit:"8478fb4fc723885b155c924d1c8c410b7a9444e6", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.8.1", GitCommit:"6af75a8fd72e2aa18a2b278cfe5c7a1c5feca7f2", GitTreeState:"clean"}

kubectl version
Client Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.0", GitCommit:"6e937839ac04a38cac63e6a7a306c5d035fe7b0a", GitTreeState:"clean", BuildDate:"2017-09-28T22:57:57Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.1", GitCommit:"3a1c9449a956b6026f075fa3134ff92f7d55f812", GitTreeState:"clean", BuildDate:"2018-01-04T11:40:06Z", GoVersion:"go1.9.2", Compiler:"gc", Platform:"linux/amd64"}

**Which chart**:
[incubator/elasticsearch] - https://github.com/helm/charts/tree/master/incubator/elasticsearch

**What happened**:
Master nodes can't discover themselves. All pods start but it seems that the discovery service isn't allowing masters to see each other. Even after a hour of them sitting they just report the following:

```
[2018-07-20T14:33:51,481][WARN ][o.e.d.z.ZenDiscovery     ] [nd-infra-elasticsearch-master-1] not enough master nodes discovered during pinging (found [[Candidate{node={nd-infra-elasticsearch-master-1}{UPiTP94YTTiWz6R0COZmKw}{LYO7sVKSRvu4hNDSD7E0zg}{10.244.2.109}{10.244.2.109:9300}, clusterStateVersion=-1}]], but needed [2]), pinging again
```

Service looks like:

```
awilhelm@MBP ~/work/k8s-config (SRE-796)*$ kubectl describe svc nd-infra-elasticsearch-discovery
Name:              nd-infra-elasticsearch-discovery
Namespace:         nd-infra
Labels:            app=elasticsearch
                   chart=elasticsearch-1.3.0
                   component=master
                   heritage=Tiller
                   release=nd-infra
Annotations:       <none>
Selector:          app=elasticsearch,component=master,release=nd-infra
Type:              ClusterIP
IP:                None
Port:              <unset>  9300/TCP
TargetPort:        %!d(string=transport)/TCP
Endpoints:         10.244.2.109:9300,10.244.7.73:9300,10.244.2.145:9300
Session Affinity:  None
Events:            <none>
```

The service responds as expected to a dig call:

```
[root@nd-infra-elasticsearch-master-0 elasticsearch]# dig nd-infra-elasticsearch-discovery.nd-infra.svc.cluster.local

; <<>> DiG 9.9.4-RedHat-9.9.4-61.el7 <<>> nd-infra-elasticsearch-discovery.nd-infra.svc.cluster.local
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 42802
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;nd-infra-elasticsearch-discovery.nd-infra.svc.cluster.local. IN	A

;; ANSWER SECTION:
nd-infra-elasticsearch-discovery.nd-infra.svc.cluster.local. 30	IN A 10.244.2.109
nd-infra-elasticsearch-discovery.nd-infra.svc.cluster.local. 30	IN A 10.244.7.73
nd-infra-elasticsearch-discovery.nd-infra.svc.cluster.local. 30	IN A 10.244.7.145

;; Query time: 2 msec
;; SERVER: 10.96.0.10#53(10.96.0.10)
;; WHEN: Fri Jul 20 14:35:28 UTC 2018
;; MSG SIZE  rcvd: 109
```

I am able to telnet to the remote pod using the IP address which proves the other master is online and listening:

```
[root@nd-infra-elasticsearch-master-0 elasticsearch]# telnet 10.244.7.73 9300
Trying 10.244.7.73...
Connected to 10.244.7.73.
Escape character is '^]'.
```

**What you expected to happen**:

The "nd-infra-elasticsearch-discovery" service should allow all masters to detect each other and resolve quorum. 

**How to reproduce it** (as minimally and precisely as possible):

helm install --name elasticsearch --namespace infra incubator/elasticsearch

kubectl log -n infra elasticsearch-master-0

**Anything else we need to know**:

### PR DESCRIPTION
**Pull request by aba182**
… name

**What this PR does / why we need it**:

Changes from global to local DNS for the elasticsearch discovery service. The fully qualified DNS entry doesn't work.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
fixes https://github.com/helm/charts/issues/6741

**Special notes for your reviewer**:
@rendhalver offered to review from the issue.

### CODE REVIEW

**Comment by rendhalver** [comment]
@aba182 Please sign the CNCF CLA as mentioned in one of the comments.
Please also increment the chart version and adjust the README.md file to reflect the new defaults.

File: incubator/elasticsearch/values.yaml [hunk 1]
```diff
@@ -10,7 +10,6 @@ image:
 
 cluster:
   name: "elasticsearch"
-  kubernetesDomain: cluster.local
```
Comment (rendhalver) [inline incubator/elasticsearch/values.yaml:13]: Does this get used anywhere else?

Comment (aba182) [reply incubator/elasticsearch/values.yaml:13]: Miss read your comment:

No I don't see it anywhere else:

```
awilhelm@MBP ~/work/charts/incubator/elasticsearch (elasticsearch_discovery)$ grep -r kubernetesDomain *
awilhelm@MBP ~/work/charts/incubator/elasticsearch (elasticsearch_discovery)$
```

**Comment by aba182** [comment]
How do we get this PR over the finish line?

**Comment by rendhalver** [comment]
I will ask for someone to look at it in Slack.

**Comment by davidkarlsen** [comment]
/lgtm cancel
Probably not due to this PR, but the tests seem to fail: https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/helm_charts/7085/pull-charts-e2e/12302/?log#log - could you look into that?

**Comment by aba182** [comment]
@davidkarlsen I don't see a failure. I do see many warnings from the client and data nodes as they wait for a master quorum. Default settings require 2 masters to have a quorum resolved before clients or data nodes are allowed to connect. 

Example can be seen for client pod "elasticsearch-v57ae9ll6i7l3zjd-client-575494b7b7-ksnq7" which goes through the following progression

These are errors when no masters have initialized:
```
I0808 18:49:28.232] [2018-08-08T18:47:55,263][WARN ][o.e.d.z.UnicastZenPing   ] [elasticsearch-v57ae9ll6i7l3zjd-client-575494b7b7-ksnq7] failed to resolve host [elasticsearch-v57ae9ll6i7l3zjd-discovery]
I0808 18:49:28.232] java.net.UnknownHostException: elasticsearch-v57ae9ll6i7l3zjd-discovery
I0808 18:49:28.233] 	at java.net.InetAddress$CachedAddresses.get(InetAddress.java:793) ~[?:?]
I0808 18:49:28.233] 	at java.net.InetAddress.getAllByName0(InetAddress.java:1495) ~[?:?]
I0808 18:49:28.233] 	at java.net.InetAddress.getAllByName(InetAddress.java:1354) ~[?:?]
I0808 18:49:28.233] 	at java.net.InetAddress.getAllByName(InetAddress.java:1288) ~[?:?]
I0808 18:49:28.233] 	at org.elasticsearch.transport.TcpTransport.parse(TcpTransport.java:933) ~[elasticsearch-6.3.1.jar:6.3.1]
I0808 18:49:28.233] 	at org.elasticsearch.transport.TcpTransport.addressesFromString(TcpTransport.java:888) ~[elasticsearch-6.3.1.jar:6.3.1]
I0808 18:49:28.233] 	at org.elasticsearch.transport.TransportService.addressesFromString(TransportService.java:710) ~[elasticsearch-6.3.1.jar:6.3.1]
I0808 18:49:28.233] 	at org.elasticsearch.discovery.zen.UnicastZenPing.lambda$resolveHostsLists$0(UnicastZenPing.java:212) ~[elasticsearch-6.3.1.jar:6.3.1]
I0808 18:49:28.233] 	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
I0808 18:49:28.233] 	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:626) [elasticsearch-6.3.1.jar:6.3.1]
I0808 18:49:28.233] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135) [?:?]
I0808 18:49:28.234] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
I0808 18:49:28.234] 	at java.lang.Thread.run(Thread.java:844) [?:?]
```

Then it progress to this type of warning once a single master is up but it requires minimum two masters:

```
I0808 18:49:28.234] [2018-08-08T18:47:58,265][WARN ][o.e.d.z.ZenDiscovery     ] [elasticsearch-v57ae9ll6i7l3zjd-client-575494b7b7-ksnq7] not enough master nodes discovered during pinging (found [[]], but needed [2]), pinging again
I0808 18:49:28.234] [2018-08-08T18:48:01,274][WARN ][o.e.d.z.ZenDiscovery     ] [elasticsearch-v57ae9ll6i7l3zjd-client-575494b7b7-ksnq7] not enough master nodes discovered during pinging (found [[Candidate{node={elasticsearch-v57ae9ll6i7l3zjd-master-0}{cRpG97ahQ3OsjxcP7LFEuw}{XteYBDGTTM-xx_LV3PB__g}{10.0.0.185}{10.0.0.185:9300}, clusterStateVersion=-1}]], but needed [2]), pinging again
```

But once enough masters are online the errors/warnings go away with lines like:

```
I0808 18:49:28.239] [2018-08-08T18:48:30,068][INFO ][o.e.c.s.ClusterApplierService] [elasticsearch-v57ae9ll6i7l3zjd-client-575494b7b7-ksnq7] detected_master {elasticsearch-v57ae9ll6i7l3zjd-master-1}{78UF-T8zRne1aqly4weADg}{eStgfsYKSzStP8EpBYdkjg}{10.0.0.187}{10.0.0.187:9300}, added {{elasticsearch-v57ae9ll6i7l3zjd-master-1}{78UF-T8zRne1aqly4weADg}{eStgfsYKSzStP8EpBYdkjg}{10.0.0.187}{10.0.0.187:9300},{elasticsearch-v57ae9ll6i7l3zjd-master-0}{cRpG97ahQ3OsjxcP7LFEuw}{XteYBDGTTM-xx_LV3PB__g}{10.0.0.185}{10.0.0.185:9300},{elasticsearch-v57ae9ll6i7l3zjd-data-0}{DyNVCpmWST2nfp1JSU7MgQ}{rjEmqDUzRmGLIq5w_D_m6w}{10.0.0.186}{10.0.0.186:9300},}, reason: apply cluster state (from master [master {elasticsearch-v57ae9ll6i7l3zjd-master-1}{78UF-T8zRne1aqly4weADg}{eStgfsYKSzStP8EpBYdkjg}{10.0.0.187}{10.0.0.187:9300} committed version [1]])
I0808 18:49:28.240] [2018-08-08T18:48:30,706][INFO ][o.e.c.s.ClusterApplierService] [elasticsearch-v57ae9ll6i7l3zjd-client-575494b7b7-ksnq7] added {{elasticsearch-v57ae9ll6i7l3zjd-client-575494b7b7-nrpgf}{6d6lfh8NQIWyofO_tAWZDQ}{zIFOpDttQECmSYTWU62xqw}{10.0.0.184}{10.0.0.184:9300},}, reason: apply cluster state (from master [master {elasticsearch-v57ae9ll6i7l3zjd-master-1}{78UF-T8zRne1aqly4weADg}{eStgfsYKSzStP8EpBYdkjg}{10.0.0.187}{10.0.0.187:9300} committed version [3]])
```

This repeats in output for both clients and data nodes until master nodes are online and allowing for connection. 

If I am miss reading the output from the test let me know and I can dig further but this is a normal state and expected. 

Thanks,

Adam

**Comment by davidkarlsen** [comment]
/lgtm
Then I was misinterpreting the logs - thanks for clarifying that!

**Comment by unguiculus** [comment]
@rendhalver @icereval OK to merge?

**Comment by rendhalver** [comment]
Maybe I need to add myself to the OWNERS for this one too.
I will stick in a PR.

**Comment by stephenmoloney** [comment]
I can't see how the cluster domain name can be wrong.

Perhaps this is closer to the core of the problem:
https://github.com/helm/charts/issues/4151#issuecomment-414565723

**Comment by lachie83** [comment]
What's the status of this?

**Comment by davidkarlsen** [comment]
@aba182 Can you merge with master && bump the version please? TIA

**Comment by aba182** [comment]
Oh goodness, I tried to follow the info given here and looks to have made a mess!!

https://github.com/helm/charts/pull/7085/checks?check_run_id=16215996

**Comment by aba182** [comment]
well damn, I am not versed this deeply in git... couldn't seem to get the rebase to be undone. Instead I have just created the following pull request which is up to date with master:

https://github.com/helm/charts/pull/7612

**Comment by aba182** [comment]
Can't seem to make git happy after the rebase. I have created a new pull off a new branch https://github.com/helm/charts/pull/7612

Going to close this PR. Sorry for the pains.

**Pull request closed without merging**